nucliadb_vectors2/src/vectors/data_point/ops_hnsw.rs

```diff
@@ -110,29 +111,29 @@ impl<'a, DR: DataRetriever> HnswOps<'a, DR> {
     }
     fn closest_up_node<L: Layer>(
         &'a self,
-        solution: &mut HashSet<Address>,
-        filters: &[&[u8]],
-        layer: L,
         x: Address,
-        y: Address,
+        query: Address,
+        layer: L,
+        filters: &[&[u8]],
+        sol_addresses: &HashSet<Address>,
+        vec_counter: &RepCounter,
     ) -> Option<(Address, f32)> {
-        solution.remove(&x);
-        let mut candidates = BinaryHeap::from([Cnx(x, self.cosine_similarity(x, y))]);
         let mut visited_nodes = HashSet::new();
+        let mut candidates = BinaryHeap::from([Cnx(x, self.cosine_similarity(x, query))]);
         loop {
             match candidates.pop() {
                 None => break None,
-                Some(Cnx(n, _))
+                Some(Cnx(n, score))
                     if !self.tracker.is_deleted(n)
-                        && !solution.contains(&n)
+                        && vec_counter.get(score) == 0
+                        && !sol_addresses.contains(&n)
                         && filters.iter().all(|label| self.tracker.has_label(n, label)) =>
```

Comment:
This predicate is a bit hard to understand at first glance. Can you add some comments or extract it in a proper function ?

Reply:
Will do

Reply:
fixed by [363880a](https://github.com/nuclia/nucliadb/pull/353/commits/363880a41a5053d81c5f3b42f5abbc68d8f4bd92)